### DISCUSSION TITLE
Application won't start

### DISCUSSION THREAD

**Discussion by mbassano** [DrewThomasson/VoxNovel#44] [Q&A]
I'm new to WSL and Ubuntu. I've followed the installation instructions for Windows 11 and did see a number of red error messages scroll by. After the installation was complete I tried to start the GUI, but I get the following error:

Starting...
Traceback (most recent call last):
  File "/home/mbassano/VoxNovel/gui_run.py", line 31, in <module>
    import pandas as pd
ModuleNotFoundError: No module named 'pandas'

The headless (whatever that is) won't start either:

Starting...
Traceback (most recent call last):
  File "/home/mbassano/VoxNovel/headless_voxnovel.py", line 58, in <module>
    import pandas as pd
ModuleNotFoundError: No module named 'pandas'

How can I fix this?

**Comment by mbassano** [comment]
OK, so I found the solution (pip install pandas) in the comments of the video. I can now start the application and select an ebook. But after a short while of processing it crashes with the following error:

Exception in Tkinter callback
Traceback (most recent call last):
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/tkinter/__init__.py", line 1921, in __call__
    return self.func(*args)
  File "/home/mbassano/VoxNovel/gui_run.py", line 438, in process_file
    booknlp = BookNLP("en", model_params)
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/site-packages/booknlp/booknlp.py", line 14, in __init__
    self.booknlp=EnglishBookNLP(model_params)
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/site-packages/booknlp/english/english_booknlp.py", line 148, in __init__
    self.entityTagger=LitBankEntityTagger(self.entityPath, tagsetPath)
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/site-packages/booknlp/english/entity_tagger.py", line 21, in __init__
    self.model.to(device)
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1174, in to
    return self._apply(convert)
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/site-packages/torch/nn/modules/module.py", line 780, in _apply
    module._apply(fn)
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/site-packages/torch/nn/modules/module.py", line 805, in _apply
    param_applied = fn(param)
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1160, in convert
    return t.to(
RuntimeError: CUDA error: the launch timed out and was terminated
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

Any help would be greatly appreciated.

**Comment by DrewThomasson** [comment]
Idk probs not enough vram

**Comment by DrewThomasson** [comment]
Try throwing this book at it idk

https://github.com/DrewThomasson/ebook2audiobookXTTS/blob/main/demo_mini_story_chapters_Drew.epub

It's a test ebook that should be able to run on the worse GPU

This'll tell us if you just don't have a good enough GPU

**Comment by mbassano** [comment]
Thanks for the quick reply!
I don't think it's GPU related, since it crashes during conversion to TXT (also with your sample ebook):

1% Converting input to HTML...
InputFormatPlugin: EPUB Input running
on /home/mbassano/VoxNovel/eBooks/demo_mini_story_chapters_Drew.epub
Found HTML cover titlepage.xhtml
Parsing all content...
34% Running transforms on e-book...
Merging user specified metadata...
Detecting structure...
        Detected chapter: Chapter 1: It Continues
        Detected chapter: Chapter 2: The Silk Road Odyssey
Flattening CSS and remapping font sizes...
Source base font size is 12.00000pt
Removing fake margins...
Cleaning up manifest...
Trimming unused files from manifest...
Creating TXT Output...
67% Running TXT Output plugin
Converting XHTML to TXT...
TXT output written to /home/mbassano/VoxNovel/eBooks/demo_mini_story_chapters_Drew.txt
Output saved to   /home/mbassano/VoxNovel/eBooks/demo_mini_story_chapters_Drew.txt
{'pipeline': 'entity,quote,supersense,event,coref', 'model': 'big'}
/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/site-packages/transformers/modeling_utils.py:463: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  return torch.load(checkpoint_file, map_location="cpu")
Exception in Tkinter callback
Traceback (most recent call last):
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/tkinter/__init__.py", line 1921, in __call__
    return self.func(*args)
  File "/home/mbassano/VoxNovel/gui_run.py", line 438, in process_file
    booknlp = BookNLP("en", model_params)
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/site-packages/booknlp/booknlp.py", line 14, in __init__
    self.booknlp=EnglishBookNLP(model_params)
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/site-packages/booknlp/english/english_booknlp.py", line 148, in __init__
    self.entityTagger=LitBankEntityTagger(self.entityPath, tagsetPath)
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/site-packages/booknlp/english/entity_tagger.py", line 21, in __init__
    self.model.to(device)
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1174, in to
    return self._apply(convert)
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/site-packages/torch/nn/modules/module.py", line 780, in _apply
    module._apply(fn)
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/site-packages/torch/nn/modules/module.py", line 805, in _apply
    param_applied = fn(param)
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1160, in convert
    return t.to(
RuntimeError: CUDA error: the launch timed out and was terminated
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

**Reply by DrewThomasson** [reply]
Still seeing GPU related issues...

Idk I'll reinstall later on my computer and see if the install is broken

**Comment by mbassano** [comment]
BTW: here's a summary of my hardware specs:

Summary
		Operating System
			Windows 11 Pro 64-bit
		CPU
			Intel Core i7 8700 @ 3.20GHz	39 °C
			Coffee Lake 14nm Technology
		RAM
			64,0GB Dual-Channel DDR4 @ 1063MHz (15-15-15-36)
		Motherboard
			ASUSTeK COMPUTER INC. WS C246M PRO Series (CPU1)	39 °C
		Graphics
			S24C450 (1920x1080@60Hz)
			LG Ultra HD (3840x2160@60Hz)
			2047MB NVIDIA GeForce GTX 950 (MSI)	57 °C
		Storage
			1863GB Hitachi HDS5C3020ALA632 (SATA )	39 °C
			11176GB TOSHIBA MG07ACA12TE (SATA )	36 °C
			476GB SK hynix BC711 HFM512GD3JX013N (Unknown (SSD))
		Optical Drives
			HL-DT-ST BD-RE BH08LS20
			TSSTcorp CDDVDW SH-224DB
		Audio
			Realtek High Definition Audio

I think my PC should be up to the task... ;-)

**Reply by DrewThomasson** [reply]
You have only 2 gb vram???

Yeha that's why Jesus

**Reply by mbassano** [reply]
Hmmm.... OK. Never had any problems, though. I regularly do HD video editing without any issues.

**Reply by DrewThomasson** [reply]
Yeah cause it's probs loading off of your regular ram


When voxnovel uses vram it ONLY uses vram

You'll have to make it so that the VoxNovel program doesn't see the GPU at the start of it

**Reply by mbassano** [reply]
Sounds complicated... 
So, what's the amount of vram you would recommend, anyway?

**Reply by DrewThomasson** [reply]
Idk try doing this


Manually launch the VoxNovel wsl env

The run these commands

The 

```bash
export CUDA_VISIBLE_DEVICES=""
```

Should make it so that your wsl stops seeing your GPU so then VoxNovel only uses the cpu

```bash
conda activate VoxNovel
cd VoxNovel/
export CUDA_VISIBLE_DEVICES=""
python gui_run.py
```

**Reply by mbassano** [reply]
I'm prorbaly doing something wrong:

PS C:\Users\mbass> wsl -d Ubuntu
(base) mbassano@Spock:/mnt/c/Users/mbass$ cd ~ && cd VoxNovel
(base) mbassano@Spock:~/VoxNovel$ Conda activate VoxNovel
Conda: command not found
(base) mbassano@Spock:~/VoxNovel$ export CUDA_VISIBLE_DEVICES=""
(base) mbassano@Spock:~/VoxNovel$ python gui_run.py
All required booknlp files are present. No action needed.
Starting...
Traceback (most recent call last):
  File "/home/mbassano/VoxNovel/gui_run.py", line 108, in <module>
    import ebooklib
ModuleNotFoundError: No module named 'ebooklib'

**Reply by DrewThomasson** [reply]
My by autocorrect on my phone captilialized conda 


The command is 

```
conda activate VoxNovel

```

**Reply by mbassano** [reply]
Right, Linux is case sensitive.... I forgot about that. Pure Windows usere here... ;-)
So, it gets a bit further now, but eventually fails with a new error, after clicking the [Continue] button on the dialog:

Rows with empty 'Text' column have been removed from Working_files/Book/book.csv
The directory Working_files/generated_audio_clips/ does not exist!
Exception in Tkinter callback
Traceback (most recent call last):
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/tkinter/__init__.py", line 1921, in __call__
    return self.func(*args)
  File "/home/mbassano/VoxNovel/speaker_editor.py", line 92, in update_combobox_colors
    self.speaker_combobox['menu'].delete(0, 'end')
  File "/home/mbassano/miniconda/envs/VoxNovel/lib/python3.10/tkinter/__init__.py", line 1681, in cget
    return self.tk.call(self._w, 'cget', '-' + key)
_tkinter.TclError: unknown option "-menu"
Traceback (most recent call last):
  File "/home/mbassano/VoxNovel/gui_run.py", line 940, in <module>
    from TTS.api import TTS
ModuleNotFoundError: No module named 'TTS'

**Reply by DrewThomasson** [reply]
Weird the auto-installer should of installed that in the conda env tho


I'll have to double check on my computer later with a new full install to see if the autoinstaller is still working

**Reply by DrewThomasson** [reply]
It look it suppose to be doing all of these pip installs idk why it doesn't seem to have them for you



```bash

    # Install necessary Python packages
    pip install styletts2 
    pip install tts==0.21.3
    pip install booknlp==1.0.7.1
    pip install bs4
    pip install -r Ubuntu_requirements.txt
    pip install ebooklib==0.18
    pip install epub2txt==0.1.6
    pip install pygame==2.6.0
    pip install moviepy==1.0.3

    # Download the spaCy language model
    pip install spacy
    python -m spacy download en_core_web_sm


```

**Reply by mbassano** [reply]
As I stated in my original post, I did see some red error messages scroll by during the installation. But since the installation simply continued, I figured they may not be important.

**Reply by DrewThomasson** [reply]
Either way this projects on hold right now

At the moment I'm slowly turning it into a full working docker image with a gradio gui instead

Then I won't have to deal with any errors like this lol

**Reply by mbassano** [reply]
I see. Makes sense. Should I just check back here for the new docker image? Or can you let me know when and where it's available? I'd love to play around with it, since I've got many eBooks.

**Reply by DrewThomasson** [reply]
I mean I guess the main question I wanna ask you cause this come up a lot is


Are you just trying to generate audiobooks?

And do you even care if it tries to give each character a different voice?

Or do you just want an audiobook read by a single narrator

**Reply by DrewThomasson** [reply]
Asking because there are a surprisingly large number of people that come to this for a normal audiobook generator

**Reply by mbassano** [reply]
Mostly I want an audiobook that sounds natural and not robotic. So far, I have no experience with similar programs. I've tried the "Evie" eBook reader app on android, but I find that too robotic with the free TTS voices. 
I do think it's very cool to have different voices for each character! Your sample just sounds fantastic! So, I'd love to try this out myself on some eBooks. But of course performance also matters. I don't know how long it would take to generate an entire novel as audiobook. If it takes several hours for each novel, I probably won't bother with it. But still I'm curious and itching to play around with it....
But is there a good free "normal" audiobook generator, that you can recommend? In my search I came across https://github.com/aedocw/epub2tts-edge, but that one lacks even the most basic installation instruction for Windows, so I don't even know how to get started with that one.... :-(

**Reply by DrewThomasson** [reply]
Great I'll just point you to all of my repos depending on your use lol 

Anyway main thing is in the meantime here's the current web-interface docker image implantations:

# For normal audiobooks where there is only one speaker (recommended less buggy):

## Ebook2audiobookxtts

[Ebook2audiobookxtts-web-demo-plz run locally this is only to show you the gui it's super slow](https://huggingface.co/spaces/drewThomasson/ebook2audiobookXTTS)

### To run locally in docker:

```bash
docker run -it -p 7860:7860 --platform=linux/amd64 registry.hf.space/drewthomasson-ebook2audiobookxtts:latest python app.py

```

## It's based on my github repo:
[Ebook2audiobookxtts](https://github.com/DrewThomasson/ebook2audiobookXTTS)



# For multiple voices (Ie give each character a different voice IE:VoxNovel:

## For a current version of auto-run VoxNovel where it selects your voices for you automatically after you give it the ebook:

*English only ebooks btw for VoxNovel*

[Auto-VoxNovel-web-demo](https://huggingface.co/spaces/drewThomasson/Auto-VoxNovel-Demo)

```bash
docker run -it -p 7860:7860 --platform=linux/amd64 registry.hf.space/drewthomasson-auto-voxnovel-demo:latest python app.py
```


## For version where you can still select the voices but it's a terminal-like only interface IE-headless

*English only ebooks btw for VoxNovel*

 
[Headless-VoxNovel-web-demo](https://huggingface.co/spaces/drewThomasson/Headless-VoxNovel-Demo)

### To run locally in docker:

```bash
docker run -it -p 7860:7860 --platform=linux/amd64 registry.hf.space/drewthomasson-headless-voxnovel-demo:latest python app.py
```

**Reply by DrewThomasson** [reply]
Keep in mind also, you only have cpu so these will all run slowly as they use XTTS


So for a 5 hour audiobook will probs takes like 8 hours to generate on your computer,

But like the outputs are great lol as seen here 

## output example from 

[Ebook2audiobookxtts](https://github.com/DrewThomasson/ebook2audiobookXTTS)

## Ebook2audiobookxtts demo 

Cloning this voice:

https://youtu.be/DkgozEpaeLw?si=RfWhk7Vm-LvHq4Dt

https://github.com/user-attachments/assets/8486603c-38b1-43ce-9639-73757dfb1031

**Reply by mbassano** [reply]
I does sound good, but 8 hours for a 5 hour audiobook...... :-( Well, I think I'll play around with it for a bit anyway. Thanks for the info!

**Reply by DrewThomasson** [reply]
Yeah I know lol

You'll have to get a nvidia GPU with at least 4gb VRAM for 5 hours generation time for a 5 hour audiobook ;/

(You can probs find one used for like 50$ online lol)



For cpu only the best option is probs this lol

https://huggingface.co/spaces/drewThomasson/ebook2audiobookpiper-tts



https://github.com/DrewThomasson/ebook2audiobookpiper-tts

**Reply by mbassano** [reply]
Thanks for the links.
Well, 5 hours is still much, much longer than I thought it would take... I was hoping something like 1 hour top per novel..... ;-) I guess I was overly optimistic about the state of the technology. Well, maybe I'll check back in a couple of years... lol.
